workflow.go

```diff
@@ -0,0 +1,165 @@
+package workflow
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// Workflow defines the trigger functions
+type Workflow struct {
+	api         string
+	accessToken string
+	eventType   string
+	client      *http.Client
+}
+
+// Params collects repository credentials and workflow information
+type Params struct {
+	BaseURL     string
+	Repo        string
+	Owner       string
+	AccessToken string
+	EventType   string
+	MaxRetries  int
+	Delay       time.Duration
+}
+
+// Configure creates an instance of Workflow
+func Configure(p Params) *Workflow {
+	host := "api.github.com"
+	if p.BaseURL != "" {
+		host = p.BaseURL
+	}
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   host,
+		Path:   fmt.Sprintf("/repos/%s/%s/dispatches", p.Owner, p.Repo),
+	}
+
+	maxRetries := 1
+	if p.MaxRetries > 1 {
+		maxRetries = p.MaxRetries
+	}
+
+	delay := 5 * time.Second
+	if p.Delay > 0*time.Second {
+		delay = p.Delay
+	}
+
+	c := http.Client{
+		Transport: &retryRoundTripper{
+			next:       http.DefaultTransport,
+			maxRetries: maxRetries,
+			delay:      delay,
+		},
+	}
+
+	return &Workflow{
+		api:         u.String(),
+		accessToken: p.AccessToken,
+		eventType:   p.EventType,
+		client:      &c,
+	}
+}
+
+func (w *Workflow) trigger(ctx context.Context) (retErr error) {
+	b, err := json.Marshal(map[string]string{
+		"event_type": w.eventType,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to marshal request body: %w", err)
+	}
+
+	body := bytes.NewReader(b)
+
+	req, err := http.NewRequestWithContext(ctx, "POST", w.api, body)
+	if err != nil {
+		return fmt.Errorf("failed to create new request: %w", err)
+	}
+
+	req.Header.Set("Accept", "application/vnd.github.v3+json")
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", fmt.Sprintf("token %s", w.accessToken))
+
+	resp, err := w.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("unexpected error triggering workflow: %w", err)
+	}
+
+	return handleResponse(resp)
+}
+
+// TriggerContext accepts context.Context and triggers a GitHub Action workflow
+func (w *Workflow) TriggerContext(ctx context.Context) error {
+	return w.trigger(ctx)
+}
+
+// Trigger triggers a GitHub Action workflow
+func (w *Workflow) Trigger() error {
+	return w.TriggerContext(context.Background())
+}
+
+type retryRoundTripper struct {
+	next       http.RoundTripper
+	maxRetries int
+	delay      time.Duration // delay between each retry
+}
+
+func (rr retryRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
```

Comment:
I see what you're trying to do here, but I think you'd be better of using a purpose built retry library and wrap your `w.trigger` in it, instead of injecting a custom Transport. I've used https://github.com/cenkalti/backoff before and it's pretty good. You can choose between constant backoff (what you have here) and an exponentially growing backoff time, which can be useful in some error scenarios.

Then you could remove the custom `retryRoundTripper` and just add this logic to `w.TriggerContext`